### PR TITLE
feat: support normalizing CRD fields that use known built-in K8S types

### LINKED ISSUE #1615: Improve diff support for resource requests/limits
The following diff is a common problem with CRDs. Argo CD reports the following as OutOfSync:

image

image

The reason is because the resources field uses IntOrString data type and what gets marshalled is inconsistent with what is in git.

We should see what we can do to auto-detect resource limits/requests equivalences and ignore these differences.

### PR DESCRIPTION
**Pull request by alexmt**
Closes #1615

PR provides generic solution for the issue described in #1615 . 

Some CRDs are re-using data structures defined in the Kubernetes source base and therefore inheriting custom JSON/YAML marshaling. Custom marshalers might serialize CRDs in a slightly different format that causes false positives during drift detection.

This is not only related to resource limits/requests. The same issue might happen with various VolumeSource ( marshaler drops `readOnly` flag if it is set to true  ) and other structs that have optional field with the `omitempty` annotation.

Implemented changes allows to provide mapping between CRD field and corresponding K8S type. E.g.

```yaml
  resource.customizations: |
    argoproj.io/Rollout:
      knownTypeFields:
      - field: spec.template.spec
        type: core/v1/PodSpec
```

As soon as mapping is configured Argo CD issues corresponding marshaler to "normalize" field value.

### CODE REVIEW

File: hack/known_types/main.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+func newCommand() *cobra.Command {
+	var (
+		docsOutputPath string = ""
+	)
+	var command = &cobra.Command{
+		Use:     "go run github.com/argoproj/argo-cd/hack/known_types ALIAS PACKAGE_PATH OUTPUT_PATH",
+		Example: "go run github.com/argoproj/argo-cd/hack/known_types corev1 k8s.io/api/core/v1 corev1_known_types.go",
+		RunE: func(c *cobra.Command, args []string) error {
+			if len(args) < 3 {
```
Comment (alexmt) [inline hack/known_types/main.go:34]: @jessesuen , @jannfis . Unfortunately, golang reflection does not allow creating instance by type name ( I spent several hours trying to find a way and pretty sure it is not supported ). So I had to use code generation. The generator is used here:

https://github.com/argoproj/argo-cd/pull/3357/files?file-filters%5B%5D=.go#diff-2389d1ede48ad7504a316ba4cf340e9dR16

If need need to support more k8s packages it should be relatively easy to do.

File: util/argo/normalizers/corev1_known_types.go [hunk 2]
```diff
@@ -0,0 +1,622 @@
+// Code generated by github.com/argoproj/argo-cd/hack/known_types. DO NOT EDIT.
```
Comment (jannfis) [inline util/argo/normalizers/corev1_known_types.go:1]: I know this is generated code, but I think we should have simple unit tests for it as well. For one, we can see whether the generated code is up-to-date with the list of known types, and also missing tests would decrease this package's coverage unnecessarily.

I think test could be as simple as the following pseudo-code:
```
knownTypes := readKnownTypesSourceFile();
for kt := range knownTypes {
  _, ok := knownTypes[kt]
  assert.True(t, ok)
}
```
What do you think?

File: docs/user-guide/diffing_known_types.txt [hunk 3]
```diff
@@ -0,0 +1,205 @@
+core/v1/AWSElasticBlockStoreVolumeSource
```
Comment (jannfis) [inline docs/user-guide/diffing_known_types.txt:1]: I'm unsure whether this file should live within docs - maybe it's better to place it in `hack` or even in `util/argo/normalizers` and point from the docs to the GH repository?

**Review by jannfis — CHANGES_REQUESTED** [review]
Wow, that's a creative approach to a complex problem @alexmt! I like it, and I agree that it could not be solved using golang's reflection - hard to do in a static typed language anyway. I just have some minor change requests, please find them below. 

Also, should we maybe incorporate the main use case (Argo Rollouts) without any further configuration? So that it works out-of-the-box?

File: pkg/apis/application/v1alpha1/types.go [hunk 4]
```diff
@@ -1081,11 +1081,17 @@ type TLSClientConfig struct {
 	CAData []byte `json:"caData,omitempty" protobuf:"bytes,5,opt,name=caData"`
 }
 
+type KnownTypeField struct {
```
[the remark this answers is not in the document]
Comment (alexmt) [reply pkg/apis/application/v1alpha1/types.go:1084]: done

File: util/argo/normalizer.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package argo
+
+import (
+	"github.com/argoproj/argo-cd/pkg/apis/application/v1alpha1"
+	"github.com/argoproj/argo-cd/util/argo/normalizers"
+	"github.com/argoproj/argo-cd/util/diff"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func NewDiffNormalizer(ignore []v1alpha1.ResourceIgnoreDifferences, overrides map[string]v1alpha1.ResourceOverride) (diff.Normalizer, error) {
```
[the remark this answers is not in the document]
Comment (alexmt) [reply util/argo/normalizer.go:10]: done

File: util/argo/normalizers/knowntypes_normalizer.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+func normalize(obj map[string]interface{}, field knownTypeField, fieldPath []string) error {
+	for i := range fieldPath {
+		if nestedField, ok, err := unstructured.NestedFieldNoCopy(obj, fieldPath[:i+1]...); err == nil && ok {
+			items, ok := nestedField.([]interface{})
+			if !ok {
+				continue
+			}
+			for j := range items {
+				item := items[j].(map[string]interface{})
```
[the remark this answers is not in the document]
Comment (alexmt) [reply util/argo/normalizers/knowntypes_normalizer.go:66]: nice catch! fixed

File: util/argo/normalizers/corev1_known_types.go (hunk 2 above)
Comment (alexmt) [reply util/argo/normalizers/corev1_known_types.go:1]: good idea. Done

File: docs/user-guide/diffing_known_types.txt (hunk 3 above)
Comment (alexmt) [reply docs/user-guide/diffing_known_types.txt:1]: moved

**Comment by alexmt** [comment]
Thank you for review @jannfis . PTAL

**Review by jannfis — APPROVED** [review]
LGTM now, but please run codegen to fix CI.

**Pull request merged**